src/com/facebook/buck/android/AndroidLibraryCompiler.java

```diff
@@ -0,0 +1,158 @@
+import java.util.Arrays;
+
+@SuppressWarnings("unused")
+abstract class AndroidLibraryCompiler implements ImplicitDepsInferringDescription<AndroidLibraryDescription.Arg> {
+
+  public abstract CompileToJarStepFactory compileToJar(AndroidLibraryDescription.Arg args, JavacOptions javacOptions, BuildRuleResolver resolver);
+
+  public Iterable<BuildRule> getDeclaredDeps(AndroidLibraryDescription.Arg args, BuildRuleResolver resolver) {
+    return ImmutableList.of();
+  }
+
+  public Iterable<BuildRule> getExtraDeps(AndroidLibraryDescription.Arg args, BuildRuleResolver resolver) {
+    return ImmutableList.of();
+  }
+
+  public boolean trackClassUsage(JavacOptions javacOptions) {
+    return javacOptions.trackClassUsage();
+  }
+
+  @Override
+  public Iterable<BuildTarget> findDepsForTargetFromConstructorArgs(
+      BuildTarget buildTarget,
+      CellPathResolver cellRoots,
+      AndroidLibraryDescription.Arg constructorArg) {
+
+    return ImmutableList.of();
+  }
+}
+
+
+class JavaLibraryCompiler extends AndroidLibraryCompiler {
+
+  @Override
+  public CompileToJarStepFactory compileToJar(
+      AndroidLibraryDescription.Arg arg,
+      JavacOptions javacOptions,
+      BuildRuleResolver resolver) {
+
+    return new JavacToJarStepFactory(javacOptions, new BootClasspathAppender());
+  }
+}
+
+
+class ScalaAndroidLibraryCompiler extends AndroidLibraryCompiler {
+
+  private final ScalaBuckConfig scalaBuckConfig;
+  private Tool scalac;
+
+  ScalaAndroidLibraryCompiler(ScalaBuckConfig config) {
+    this.scalaBuckConfig = config;
+  }
+
+  private Tool getScalac(BuildRuleResolver resolver) {
+    if (scalac == null) {
+      scalac = scalaBuckConfig.getScalac(resolver);
+    }
+    return scalac;
+  }
+
+  @Override
+  public boolean trackClassUsage(JavacOptions javacOptions) {
+    return false;
+  }
+
+  @Override
+  public Iterable<BuildRule> getDeclaredDeps(AndroidLibraryDescription.Arg arg, BuildRuleResolver resolver) {
+    return ImmutableList.of(resolver.getRule(scalaBuckConfig.getScalaLibraryTarget()));
+  }
+
+  @Override
+  public Iterable<BuildRule> getExtraDeps(AndroidLibraryDescription.Arg arg, BuildRuleResolver resolver) {
+    return getScalac(resolver).getDeps(new SourcePathResolver(resolver));
+  }
+
+  @Override
+  public CompileToJarStepFactory compileToJar(AndroidLibraryDescription.Arg arg, JavacOptions javacOptions, BuildRuleResolver resolver) {
+    return new ScalacToJarStepFactory(
+        getScalac(resolver),
+        ImmutableList.<String>builder()
+            .addAll(scalaBuckConfig.getCompilerFlags())
+            .addAll(arg.extraArguments.get())
+            .addAll(Iterables.transform(
+                resolver.getAllRules(scalaBuckConfig.getCompilerPlugins()),
+                new Function<BuildRule, String>() {
+                  @Override
+                  public String apply(BuildRule input) {
+                    return "-Xplugin:" + input.getPathToOutput();
+                  }
+                }))
```

Comment:
To achieve better reuse it could be good to replace `AndroidLibraryCompiler` with common `JvmLibraryCompiler`, this would be more general compiler provider without references to android (placed in `jvm.common`), then `ScalaLIbraryDescription` (and others) would use it the same way `AndroidLibraryDescription` does.
I'm just afraid that this will end up being a bigger refactoring, and I'm not sure if I want to do it as a part of this PR.